### PR TITLE
Add year_of_week Spark function

### PR DESCRIPTION
**Pull request by PHILO-HE**
This is mostly copied from presto [YearOfWeek function](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DateTimeFunctions.h#L594). The differences are:
1) Spark only supports Date type input.
2) The result type is int64_t for Presto, but int32_t for Spark.

See doc: https://docs.databricks.com/en/sql/language-manual/functions/extract.html

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@rui-mo, could you review this pr? Thanks!

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 1]
```diff
@@ -968,5 +968,21 @@ TEST_F(DateTimeFunctionsTest, makeYMInterval) {
       fromYear(-178956971), "Integer overflow in make_ym_interval(-178956971)");
 }
 
+TEST_F(DateTimeFunctionsTest, yearOfWeek) {
+  const auto yearOfWeek = [&](std::optional<int32_t> date) {
+    return evaluateOnce<int64_t, int32_t>("year_of_week(c0)", {date}, {DATE()});
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:973]: int64_t -> int32_t?

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 2]
```diff
@@ -98,6 +98,36 @@ struct WeekFunction : public InitSessionTimezone<T> {
...
+    // in the same week as the next January 1 and this January 1 is a Thursday
+    // or before.
+    if (UNLIKELY(
+            dateTime.tm_mon == 11 && dateTime.tm_mday >= 29 &&
+            dateTime.tm_mday - isoWeekDay >= 31 - 3)) {
+      return 1900 + dateTime.tm_year + 1;
+    }
+    // The first few days in January may belong to the last year if they are
+    // in the same week as January 1 and January 1 is a Friday or after.
+    else if (UNLIKELY(
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:117]: nit: no need to use else after return.

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 3]
```diff
@@ -98,6 +98,36 @@ struct WeekFunction : public InitSessionTimezone<T> {
...
+                 dateTime.tm_mon == 0 && dateTime.tm_mday <= 3 &&
+                 isoWeekDay - (dateTime.tm_mday - 1) >= 5)) {
+      return 1900 + dateTime.tm_year - 1;
+    } else {
+      return 1900 + dateTime.tm_year;
```
Comment (rui-mo) [inline velox/functions/sparksql/DateTimeFunctions.h:122]: Ditto.

**Review by rui-mo — COMMENTED** [review]
Thanks.

**Comment by PHILO-HE** [comment]
Hi @mbasmanova, Rui has no more comment. Could you spare some time to review this PR? Thanks!

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 4]
```diff
@@ -98,6 +98,35 @@ struct WeekFunction : public InitSessionTimezone<T> {
   }
 };
 
+template <typename T>
+struct YearOfWeekSparkFunction : public InitSessionTimezone<T> {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/DateTimeFunctions.h:102]: Why is it `YearOfWeekSparkFunction`? The namespace is enough to distinguish them.

Comment (mbasmanova) [reply velox/functions/sparksql/DateTimeFunctions.h:102]: That's a good point. @PHILO-HE Would you take a look?

Comment (PHILO-HE) [reply velox/functions/sparksql/DateTimeFunctions.h]: @jinchengchenghh, thanks for your comment! Just fixed. If you have more comment, please kindly let me know.

**Comment by PHILO-HE** [comment]
@mbasmanova, could you review this pr? Thanks!

**Comment by PHILO-HE** [comment]
Hi @mbasmanova, could you spare some time to review this small patch?

**Comment by mbasmanova** [comment]
@PHILO-HE There are CI failures. Would you rebase to see if they go away?

CC: @kgpai

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 5]
```diff
@@ -98,6 +98,35 @@ struct WeekFunction : public InitSessionTimezone<T> {
   }
 };
 
+template <typename T>
+struct YearOfWeekFunction : public InitSessionTimezone<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE int32_t computeYearOfWeek(const std::tm& dateTime) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/DateTimeFunctions.h:105]: this function is used only once in a method that doesn't have anything else; let's fold it into the caller.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 6]
```diff
@@ -968,5 +968,21 @@ TEST_F(DateTimeFunctionsTest, makeYMInterval) {
       fromYear(-178956971), "Integer overflow in make_ym_interval(-178956971)");
 }
 
+TEST_F(DateTimeFunctionsTest, yearOfWeek) {
+  const auto yearOfWeek = [&](std::optional<int32_t> date) {
+    return evaluateOnce<int32_t, int32_t>("year_of_week(c0)", {date}, {DATE()});
+  };
+  EXPECT_EQ(std::nullopt, yearOfWeek(std::nullopt));
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:975]: No need to tests nulls. Let's remove.

**Review by mbasmanova — COMMENTED** [review]
@PHILO-HE Looks good % a nit.

**Comment by PHILO-HE** [comment]
@mbasmanova, I have fixed all comments, and all checks have passed.

**Review by mbasmanova — APPROVED** [review]
@PHILO-HE Looks good to me. Thanks.

**Pull request closed without merging**